sklearn/metrics/_pairwise_distances_reduction/_argkmin_lrd.pyx.tp

```diff
@@ -0,0 +1,125 @@
+        intp_t[:, ::1] actual_indices
+        intp_t factor
+
+    @classmethod
+    def compute(
+        cls,
+        X,
+        Y,
+        intp_t k,
+        str metric="euclidean",
+        chunk_size=None,
+        dict metric_kwargs=None,
+        str strategy=None,
+        bint train=False,
+        bint return_distance=False,
+    ):
+        # Use a generic implementation that handles most scipy
+        # metrics by computing the distances between 2 vectors at a time.
+        pda = ArgKminLRD{{name_suffix}}(
+            datasets_pair=DatasetsPair{{name_suffix}}.get_for(X, Y, metric, metric_kwargs),
+            k=k,
+            chunk_size=chunk_size,
+            strategy=strategy,
+            train=train,
+        )
+
+        # Limit the number of threads in second level of nested parallelism for BLAS
+        # to avoid threads over-subscription (in GEMM for instance).
+        with threadpool_limits(limits=1, user_api="blas"):
+            if pda.execute_in_parallel_on_Y:
+                pda._parallel_on_Y()
+            else:
+                pda._parallel_on_X()
+
+        return pda._finalize_results(return_distance)
+
+    def __init__(
+        self,
+        DatasetsPair{{name_suffix}} datasets_pair,
+        chunk_size=None,
+        strategy=None,
+        intp_t k=1,
+        bint train=False,
+    ):
+        super().__init__(
+            datasets_pair=datasets_pair,
+            chunk_size=chunk_size,
+            strategy=strategy,
+            k=k,
+        )
+
+        self.lrd = np.zeros(self.n_samples_X, dtype=np.float64)
+        self.factor = 1 if train else 0
+        self.actual_distances = np.zeros(
+            (self.n_samples_X, self.k - self.factor), dtype=np.float64
+        )
+        self.actual_indices = np.zeros(
+            (self.n_samples_X, self.k - self.factor), dtype=np.intp
+        )
+
+    cdef inline void _compute_local_reachability_density(
+        self,
+        bint return_distance,
+    ) noexcept nogil:
+        cdef:
+            intp_t sample_index, neighbor_rank, neighbor_idx
+            float64_t dist_sample, dist_neighbor, dist_reach
+            float64_t overall_reachability
+            float64_t[:, ::1] distances = self.argkmin_distances
+            intp_t[:, ::1] indices = self.argkmin_indices
+
+        for sample_index in prange(self.n_samples_X, schedule='static', nogil=True,
+                                   num_threads=self.effective_n_threads):
+            overall_reachability = 0
+            for neighbor_rank in range(self.factor, self.k):
+                neighbor_idx = indices[sample_index, neighbor_rank]
+                dist_sample = self.datasets_pair.distance_metric._rdist_to_dist(
+                    fmax(distances[sample_index, neighbor_rank], 0.0)
+                )
+                dist_neighbor = self.datasets_pair.distance_metric._rdist_to_dist(
+                    fmax(distances[neighbor_idx, self.k - 1], 0.0)
+                )
```

Comment:
1. I think this is how it is done in the base implementation of argkmin.
https://github.com/OmarManzoor/scikit-learn/blob/a9b06a7cbfe3a4f43534a1a840231eb09ec8fb69/sklearn/metrics/_pairwise_distances_reduction/_argkmin.pyx.tp#L285-L287

Reply:
2. I agree with this point but for some reason when I first use the original distances and then compute the max and then apply rdist_to_dist on the resulting value some tests fail possibly because of loss of precision.

Reply:
For point 2 I realized that we also need to convert to the actual distance in this line 
```python
if return_distance:
                    self.actual_distances[sample_index, neighbor_rank - self.factor] = dist_sample
```
So we would be making two calls to self.datasets_pair.distance_metric._rdist_to_dist anyways. I think in that case the original implementation looks more clear.

Reply:
I still wonder regarding the use of `_rdist_to_dist`. It is fine if we do it in the `if return_distances` branch, since at worst it'll still be as many computations as originally, but if `return_distances=False` we'll save on one.

With that being said, I don't really know how much of a performance difference it would make. If you could get it working and demonstrate that it _is not_ a significant speedup when `return_distances=False` then we can keep it as-is.

The failing tests in such a case though is also curious, and perhaps should be looked at.

This all is not a high priority though, mainly a small curiosity.

Reply:
I think I agree with you regarding the condition for return_distances. Currently however since we are handling the fit method we will always return distances as that is a requirement.

Reply:
As for the failing tests that was probably because I was not using _rdist_to_dist when setting the actual distances.